ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -2650,11 +2663,15 @@ private StoreFindToken revalidateStoreFindToken(StoreFindToken token,
       case IndexBased:
         // An index based token, but the offset is not in the segments, might be caused by the compaction
         if (!indexSegments.containsKey(offset)) {
-          if (!ignoreResetKey && token.getResetKey() != null) {
-            // reset key is present in the token
-            revalidatedToken = generateTokenBasedOnResetKey(token, indexSegments);
+          if (shouldRebuildTokenBasedOnCompactionHistory && !sanityCheckFailed) {
+            revalidatedToken = generateTokenBasedOnIndexSegmentOffset(token, indexSegments);
           } else {
-            revalidatedToken = new StoreFindToken();
+            if (!ignoreResetKey && token.getResetKey() != null) {
+              // reset key is present in the token
+              revalidatedToken = generateTokenBasedOnResetKey(token, indexSegments);
+            } else {
+              revalidatedToken = new StoreFindToken();
```

Comment:
Do we want a metric to calculate how much time it has been reduced for the token to be reset to the beginning of legSegments?

Reply:
I don't understand, what do we want to record here? The time we to generate a new token?

Reply:
Nvm, we can monitoring the replication lag.